### PR TITLE
Unexpected error code for panics in middleware

### PR DESCRIPTION
**Pull request by eesmerdag**
https://www.notion.so/figmentnetworks/Middleware-should-return-http-status-500-when-there-is-a-panic-8fd60a23dde54de3b706613d9aea1ab0

### CODE REVIEW

File: server/middleware.go [hunk 1]
```diff
@@ -29,6 +30,7 @@ func MetricMiddleware() gin.HandlerFunc {
 func ErrorReportingMiddleware() gin.HandlerFunc {
 	return func(c *gin.Context) {
 		defer reporting.RecoverError()
+		http.ServerError(c, errors.New("Unexpected internal server error occurred"))
```
Comment (silentlight) [inline server/middleware.go:33]: Nice work @eesmerdag! You can create a var with ErrUnexpectedError on top and assign that error to it. It is always good to have error on top ;)

**Comment by lukanus** [comment]
Maybe I'm not familiar with gin, but shouldn't this be on the level of application code / business logic itself? There are different scenarios where you may like to return different errors, with different error messages (Unexpected internal server error occurred) doesn't tell us much.) So open question, should that be on the application, and not middleware layer ?

File: server/middleware.go (hunk 1 above)
Comment (eesmerdag) [reply server/middleware.go]: :thumbsup: Let's decide this is going to be our standard then?

**Comment by eesmerdag** [comment]
> Maybe I'm not familiar with gin, but shouldn't this be on the level of application code / business logic itself? There are different scenarios where you may like to return different errors, with different error messages (Unexpected internal server error occurred) doesn't tell us much.) So open question, should that be on the application, and not middleware layer ?

At first, I thought that the recovery should be on each handler. But this is fine I guess🤔 I can change the error message but actually error message goes to rollbar. this is just a message from Api 🤔🤔🤔🤔🤔🤔🤔🤔

File: server/middleware.go (hunk 1 above)
Comment (eesmerdag) [reply server/middleware.go]: I used goroutine not to lose message on rollbar. Now it is fixed

**Comment by lukanus** [comment]
I believe that any panic-based error in go is wrong and should be removed as soon as possible :) 
So instead of doing this just return this error from the place it happened. 
If something panics - solve the panic ( guard the code to prevent that ) instead, and then return 5XX/4XX directly from endpoint :)

**Comment by eesmerdag** [comment]
> I believe that any panic-based error in go is wrong and should be removed as soon as possible :)
> So instead of doing this just return this error from the place it happened.
> If something panics - solve the panic ( guard the code to prevent that ) instead, and then return 5XX/4XX directly from endpoint :)

This code piece takes only unhandled errors that may make the application stop. So this is something we do not expect but possible bug on running time. If i knew for which case it throw panic, i would handle it as error. But for this task, it is not the case.  And it returns http code 200 for now on prod whereas it should actually return 500. This is only what implemented 🤔🤔🤔🤔🤔🤔 If you still do not aggree, please correct me with an example since it is too late for me 😆😆😆😆😆

**Comment by lukanus** [comment]
Is this panic reproducible? Or like - we knew in which case the problem existed? or this is just nice to have?

**Comment by eesmerdag** [comment]
> Is this panic reproducible? Or like - we knew in which case the problem existed? or this is just nice to have?

I can trigger the panic on my local 😋  Hopefully we wont have any panic case in prod. And yes this is nice to have + should be this way

**Comment by silentlight** [comment]
Yeah we always return errors instaed of panicing, so as Emir noted it is just in case piece of code

**Comment by eesmerdag** [comment]
@silentlight @lukanus could you approve this if all looking good to you?

**Comment by lukanus** [comment]
@eesmerdag I've just seen the latest version. Are you sure about. go keyword there ? It seems it suppose to fail/abort gin's internal context https://github.com/figment-networks/polkadothub-indexer/blob/dev/usecase/http/responders.go#L47 Running that in goroutine before moving forward would apply this in random moment, if - at all. Plus there is a risk that c *gin.Context would be nil at that point. It seems that this should be a normal function call.

**Comment by eesmerdag** [comment]
> @eesmerdag I've just seen the latest version. Are you sure about. go keyword there ? It seems it suppose to fail/abort gin's internal context https://github.com/figment-networks/polkadothub-indexer/blob/dev/usecase/http/responders.go#L47 Running that in goroutine before moving forward would apply this in random moment, if - at all. Plus there is a risk that c *gin.Context would be nil at that point. It seems that this should be a normal function call.

I am sure about that go keyword for goroutine. But i could not replicate that case when Context is nil 🤔🤔🤔🤔🤔🤔🤔🤔🤔

**Comment by lukanus** [comment]
Just saying that is should not be a goroutine :) It should just run normally before c.Next()

**Comment by eesmerdag** [comment]
> Just saying that is should not be a goroutine :) It should just run normally before c.Next()

Then in that case, the logs won't go to rollbar. Actually it does not reach "reporting.RecoverError()", that is why i had to use goroutine

**Comment by lukanus** [comment]
> Then in that case, the logs won't go to rollbar. Actually it does not reach "reporting.RecoverError()", that is why i had to use goroutine
go routine run is always planned for "some unknown later time in future". Meanwhile in this particular place we need to apply this before running c.Next() so it will fail the request properly. 
If defer is never called, ( I don't really know Gin)  I would say it's is because this function is constantly looping in this line.
https://github.com/gin-gonic/gin/blob/master/context.go#L164
So when you apply Abort() - you should not be calling `c.Next()`

**Comment by eesmerdag** [comment]
> > Then in that case, the logs won't go to rollbar. Actually it does not reach "reporting.RecoverError()", that is why i had to use goroutine
> 
> go routine run is always planned for "some unknown later time in future". Meanwhile in this particular place we need to apply this before running c.Next() so it will fail the request properly.
> If defer is never called, ( I don't really know Gin) I would say it's is because this function is constantly looping in this line.
> https://github.com/gin-gonic/gin/blob/master/context.go#L164
> So when you apply Abort() - you should not be calling `c.Next()`

hmmm so the context of goroutine does not fit to this case, Do you recommend waitGroup in this case?

**Comment by lukanus** [comment]
@eesmerdag waitGroup? this is all synchronous, why would you need a wait group here ?

**Review by silentlight — APPROVED** [review]
Bit complicated but I think it is ok. Have you tested it thoroughly?

File: server/server.go [hunk 2]
```diff
@@ -1,29 +1,43 @@
...
 
 // New returns a new server instance
-func New(cfg *config.Config, handlers *usecase.HttpHandlers) *Server {
+func New(cfg *config.Config, handlers *usecase.HttpHandlers) (*Server, error) {
+	config := zap.NewDevelopmentConfig()
+	logger, error := config.Build()
+	if error != nil {
+		return nil, error
+	}
+
```
Comment (lukanus) [inline server/server.go:33]: You don't need this part it is already initialized in shared package. 
"github.com/figment-networks/polkadothub-indexer/utils/logger" 
So below you can assign just `logger.Logger`

File: server/server.go [hunk 3]
```diff
@@ -49,3 +63,9 @@ func (s *Server) startMetricsServer() error {
 	return metric.NewServerMetric().StartServer(s.cfg.ServerMetricAddr, s.cfg.MetricServerUrl)
 }
 
+func (s *Writer) Write(p []byte) (n int, err error) {
+	err = errors.New(string(p))
+	s.logger.Error(err.Error())
```
Comment (lukanus) [inline server/server.go:68]: this would make use of `defer s.logger.Sync()` to guarantee that this write will happen :)

**Review by lukanus — COMMENTED** [review]
I suppose I am literally Hitler at that point :) But you don't need to initialize zap.
 @silentlight has already done it,  so you can just link the existing

File: server/server.go (hunk 2 above)
Comment (silentlight) [reply server/server.go:33]: Right, that's good catch

**Comment by eesmerdag** [comment]
> Bit complicated but I think it is ok. Have you tested it thoroughly?

I triggered the panic case by changing the code. Is there any other case to test?

**Review by lukanus — APPROVED** [review]
Looks good!

**Pull request merged**